README.md
# 📸 InstaHashtag

<p align="center">

  <a href="https://github.com/column-street/InstaHashtag/actions?query=workflow%3ABuild">
    <img src="https://github.com/column-street/InstaHashtag/workflows/Build/badge.svg?branch=master">
  </a>

  <a href="https://column-street.github.io/InstaHashtag/">
    <img src="https://github.com/column-street/InstaHashtag/workflows/Docs/badge.svg?branch=master">
  </a>

  <a href="https://coveralls.io/github/column-street/InstaHashtag?branch=master">
    <img src="https://coveralls.io/repos/github/column-street/InstaHashtag/badge.svg?branch=master&service=github">
  </a>

  <a href="https://opensource.org/licenses/MIT">
    <img src="https://img.shields.io/badge/License-MIT-yellow.svg">
  </a>
</p>


Unofficial Python API wrapper for [DisplayPurposes](https://displaypurposes.com/).

## Installing

```
pip install instahashtag
```

## Documentation

Documentation for this module can be found [here](https://column-street.github.io/InstaHashtag/).

## Use

This library is built as a layer-based module and may be used in different ways depending on the use 
case of the user. In this README file only the `wrapper` module is documented, but for further
use cases please check out the docs above.

### Wrapper

A complete Python wrapper around the API.

```python
from instahashtag import Tag, Graph, Maps

tag = Tag(hashtag="instagram")
graph = Graph(hashtag="instagram")
maps = Maps(
    x1=-80.48712034709753,
    y1=25.750749758162012,
    x2=-79.82794065959753,
    y2=25.854604964203453,
    zoom=12,
)
```

Check out the documentation above to understand how the objects behave, and what attributes are accessible.


ChangeLog
CHANGES
=======

* Added Github action
* Added tests
* Added documentation
* Added project


tests/test_map.py
import pytest
from instahashtag import Maps


class Test_Tag_sync:
    def test_init(self):
        maps = Maps(
            x1=-80.48712034709753,
            y1=25.750749758162012,
            x2=-79.82794065959753,
            y2=25.854604964203453,
            zoom=12,
        )

    def test_init_nonexisting_map(self):
        maps = Maps(
            x1=0,
            y1=0,
            x2=1,
            y2=1,
            zoom=12,
        )

    def test_tags_gt(self):
        maps = Maps(
            x1=-80.48712034709753,
            y1=25.750749758162012,
            x2=-79.82794065959753,
            y2=25.854604964203453,
            zoom=12,
        )
        largest = max(maps.tags)


class Test_Tag_async:
    @pytest.mark.asyncio
    async def test_init(self):
        maps = Maps(
            x1=-80.48712034709753,
            y1=25.750749758162012,
            x2=-79.82794065959753,
            y2=25.854604964203453,
            zoom=12,
            aio=True,
        )
        await maps.call()


tests/test_tag.py
import pytest
from instahashtag import Tag


class Test_Tag_sync:
    def test_init(self):
        tag = Tag("miami")

    def test_init_nonexisting_hashtag(self):
        tag = Tag("random_tag_that_does_not_exist")

    def test_tags_gt(self):
        tag = Tag("miami")
        largest = max(tag.results)

class Test_Tag_async:
    @pytest.mark.asyncio
    async def test_init(self):
        tag = Tag("miami", aio=True)
        await tag.call()


tests/test_graph.py
import pytest
from instahashtag import Graph


class Test_Graph_sync:
    def test_init(self):
        graph = Graph("miami")

    def test_init_nonexisting_hashtag(self):
        graph = Graph("random_tag_that_does_not_exist")


class Test_Graph_async:
    @pytest.mark.asyncio
    async def test_init(self):
        graph = Graph("miami", aio=True)
        await graph.call()


instahashtag/api.py
from typing import Union, Awaitable

from . import http


def tag(hashtag: str, aio: bool = False) -> Union[dict, Awaitable]:
    """Sends a request to the server.

    Args:
        hashtag: Hashtag to retrieve info from.
        aio: If set to ``True`` will return a Future for use with ``await`` keyword. Defaults to ``False``.

    Return:
        Dictionary with given data (see Notes below).

    Note:
        The return dictionary follows the following format:

        .. code-block:: python

            {
                geo: [float, float],
                rank: int,
                results: [
                    {
                        absRelevance: float,
                        geo: [float, float],
                        media_count: int,
                        rank: int,
                        relevance: int,
                        tag: str
                    }
                    ...
                ],
                tag: str,
                tagExists: bool
            }
    """

    if aio:
        return http.Aiohttp.tag(hashtag=hashtag)
    else:
        return http.Requests.tag(hashtag=hashtag)


def graph(hashtag: str, aio: bool = False) -> Union[dict, Awaitable]:
    """Generates graphing query to send to the server.

    Args:
        hashtag: Hashtag to retrieve info from.
        aio: If set to ``True`` will return a Future for use with ``await`` keyword. Defaults to ``False``.

    Return:
        Dictionary with given data (see Notes below).

    Note:
        The return dictionary follows the following format:

        .. code-block:: python

            {
                edges: [
                    {
                        a: str,
                        b: str,
                        id: str,
                        weight: float
                    }
                    ...
                ],
                exists: bool,
                nodes: [
                    id: str,
                    relevance: float,
                    weight: float,
                    x: float,
                    y: float
                ],
                query: string,
                root_pos: [float, float]
            }
    """

    if aio:
        return http.Aiohttp.graph(hashtag=hashtag)
    else:
        return http.Requests.graph(hashtag=hashtag)


def maps(
    x1: float,
    y1: float,
    x2: float,
    y2: float,
    zoom: float = 1,
    aio: bool = False,
) -> Union[dict, Awaitable]:
    """Generates graphing query to send to the server.

    Args:
        x1: Top left x-coordinate corner of the map.
        y1: Top left y-coordinate corner of the map.
        x2: Bottom right x-coordinate corner of the map.
        y2: Bottom right y-coordinate corner of the map.
        zoom: Number from 2 to 16 that designates the zoom factor.
        aio: If set to ``True`` will return a Future for use with ``await`` keyword. Defaults to ``False``.

    Return:
        Dictionary with given data (see Notes below).

    Note:
        The return dictionary follows the following format:

        .. code-block:: python

            {
                count: int
                tags: [
                    {
                        "centroid": [
                            0: float
                            1: float
                        ],
                        tag: str,
                        weight: int
                    }
                    ...
                ]
            }
    """

    if aio:
        return http.Aiohttp.maps(x1=x1, y1=y1, x2=x2, y2=y2, zoom=zoom)
    else:
        return http.Requests.maps(x1=x1, y1=y1, x2=x2, y2=y2, zoom=zoom)


instahashtag/http.py
import json
from abc import ABC, abstractmethod
from typing import Any

import aiohttp as aiohttp_module
import requests as requests_module

from . import utils


class endpoints:
    """API endpoints."""

    tag = "https://apidisplaypurposes.com/tag/{}"
    graph = "https://apidisplaypurposes.com/graph/{}"
    maps = "https://apidisplaypurposes.com/local/?bbox={},{},{},{}&zoom={}"


class Base(ABC):
    """Base class that properly processes and calls the API.

    Allows one to "plug-and-play" with other request libraries with ease. If one wants to,
    they may inherit from this class and overwrite the abstract :py:func:`call` method.

    Example
        While the module ``httpx`` is not supported out of the box, one may inherit from the
        :py:class:`Base` class and utilize it.

        .. code-block:: python

            from instahashtag.http import Base
            import httpx

            class httpx_io(Base):
                def call(self):
                    req = httpx.get(self.endpoint, headers=self.headers)
                    resp = req.text

                    # Note that self.process is just a simple wrapper for 'json.loads'.
                    # One has the option to also overwrite this to catch any extra error(s),
                    # or run processes on the data.
                    return self.process(resp)

            class httpx_aio(Base):
                async def call(self):
                    async with httpx.AsyncClient() as client:
                        req = await client.get(self.endpoint, headers=self.headers)

                    resp = req.text
                    return self.process(resp)

            def io():
                tag = httpx_io.tag(hashtag="instagram")
                graph = httpx_io.graph(hashtag="instagram")
                maps = httpx_io.maps(
                    x1=-80.48712034709753,
                    y1=25.750749758162012,
                    x2=-79.82794065959753,
                    y2=25.854604964203453,
                    zoom=12,
                )

            def aio():
                tag = await httpx_io.tag(hashtag="instagram")
                graph = await httpx_io.graph(hashtag="instagram")
                maps = await httpx_io.maps(
                    x1=-80.48712034709753,
                    y1=25.750749758162012,
                    x2=-79.82794065959753,
                    y2=25.854604964203453,
                    zoom=12,
                )

    """

    def __init__(self, endpoint: str, headers: dict) -> None:
        self.endpoint = endpoint
        self.headers = headers

    @abstractmethod
    def call(self) -> NotImplemented:  # pragma: no cover
        """Abstract method that needs to be written to query the API endpoint."""
        raise NotImplementedError

    @staticmethod
    def process(resp: str) -> dict:
        """Processes the reply returned back by :py:data:`tag`, :py:data:`graph`, and :py:data:`maps`.

        The current implementation of this function is a simple ``json.loads(resp)``, returning
        back a Python dictionary object. However, one may chose to overwrite this function to
        process the data in some more meaningful way.
        """
        return json.loads(resp)

    @classmethod
    def tag(cls, hashtag: str) -> Any:
        """Sends an API request to the ``tag`` endpoint."""

        endpoint = endpoints.tag.format(hashtag)
        headers = utils.generate_header(hashtag=hashtag)

        obj = cls(endpoint, headers)
        return obj.call()

    @classmethod
    def graph(cls, hashtag: str) -> Any:
        """Sends an API request to the ``graph`` endpoint."""

        endpoint = endpoints.graph.format(hashtag)
        headers = utils.generate_header(hashtag=hashtag)

        obj = cls(endpoint, headers)
        return obj.call()

    @classmethod
    def maps(cls, x1: float, y1: float, x2: float, y2: float, zoom: int) -> Any:
        """Sends an API request to the ``maps`` endpoint."""

        endpoint = endpoints.maps.format(x1, y1, x2, y2, zoom)
        headers = utils.generate_header(hashtag=None)

        obj = cls(endpoint, headers)
        return obj.call()


class Requests(Base):
    """Class that utilitizes the ``request`` module to make API request."""

    def call(self):
        req = requests_module.get(url=self.endpoint, headers=self.headers)
        resp = req.text

        return self.process(resp)


class Aiohttp(Base):
    """Class that utilitizes the ``aiohttp`` module to make API request."""

    async def call(self):
        async with aiohttp_module.ClientSession(headers=self.headers) as session:
            async with session.get(self.endpoint) as response:
                resp = await response.text()

        return self.process(resp)


instahashtag/utils.py
import hashlib

USERAGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.146 Safari/537.36"
STRING = 'function(d){var r = M(V(Y(X(d),8*d.length)));return r.toLowerCase()};function M(d){for(var _,m="0123456789ABCDEF",f="",r=0;r<d.length;r++)_=d.charCodeAt(r)'
HEADERS = {
    "authority": "apidisplaypurposes.com",
    "user-agent": USERAGENT,
    "dnt": "1",
    "api-token": None,
    "accept": "*/*",
    "origin": "https://displaypurposes.com",
    "sec-fetch-site": "cross-site",
    "sec-fetch-mode": "cors",
    "sec-fetch-dest": "empty",
    "referer": "https://displaypurposes.com/",
    "accept-language": "en-US,en;q=0.9",
}


def generate_token(hashtag: str) -> str:
    """Generates the required 'api-token' to send request to DisplayPurposes.

    Args:
        tag: Hashtag to generate token for.

    Returns:
        dict: Dictionary representation of the request headers.
    """
    value = "{}|{}|{}".format(USERAGENT, hashtag, STRING).encode("utf-8")
    md5 = hashlib.md5(value)
    return md5.hexdigest()


def generate_header(hashtag: str = None) -> dict:
    """Generates specific headers that DisplayPurposes requires.

    Args:
        tag: Hashtag to generate headers for.

    Returns:
        dict: Dictionary representation of the request headers.

    Note:
        'map' request does not use a generated 'api-token'
        header, and instead uses the string "test".
    """
    headers = HEADERS.copy()

    if hashtag:
        headers["api-token"] = generate_token(hashtag)
    else:
        headers["api-token"] = "test"

    return headers


instahashtag/wrapper/tag.py
from typing import List

from .. import api


class TagResult:
    def __init__(
        self,
        tag: str,
        rank: int,
        geo: List[float],
        media_count: int,
        relevance: int,
        absRelevance: float,
    ) -> None:
        """Object that represents the individual items inside the ``Tag.results`` list.

        Attributes:
            tag: Related hashtag.

                .. code-block:: python

                    result.tag # >>> miamibeach

            rank: Ranking of the tag.

                .. code-block:: python

                    result.rank # >>> 74

            geo: List of floats that contains coordinates to the hashtag.

                .. code-block:: python

                    result.geo # >>> [25.819434533299013, -80.16981253812398]

            media_count: Number of posts in the hashtag.

                .. code-block:: python

                    result.media_count # >>> 4329283

            relevance: Relevance of the hashtag to the queried hashtag.

                .. code-block:: python

                    result.relevance # >>> 99

            absRelevance: Absolute relevance to the queried hashtag.

                .. code-block:: python

                    result.absRelevance # >>> 0.0060874452062492975
        """
        self.tag = tag
        self.rank = rank
        self.geo = geo
        self.media_count = media_count
        self.relevance = relevance
        self.absRelevance = absRelevance

    def __gt__(self, other: "Result") -> bool:
        """Allows comparisson of :py:class:`Result` objects by their ``rank``.

        Note:
            This function allows the user to utilize the Python built-in functions.

            .. code-block:: python

                from instahashtag import Tag

                tag = Tag("miami")

                min(tag.results)
                max(tag.results)
                sort(tag.results)

        """
        return self.rank > other.rank

    def __repr__(self) -> str: # pragma: no cover
        return "Result(tag={}, rank={}, geo={}, media_count={}, relevance={}, absRelevance={})".format(
            self.tag,
            self.rank,
            self.geo,
            self.media_count,
            self.relevance,
            self.absRelevance,
        )

    def __str__(self) -> str: # pragma: no cover
        return self.__repr__()


class Tag:
    def __init__(self, hashtag: str, aio: bool = False) -> None:
        """Initializes a new Tag object.

        .. code-block:: python

            from instahashtag import Tag

            tag = Tag("miami")

        Attributes:
            hashtag: Hashtag that was used to retrieve information from.

               .. code-block:: python

                    tag.hashtag # >>> miami

            geo: List containing two float coordinates relating to the hashtag.

                .. code-block:: python

                    tag.geo # >>> [25.821117872941034, -80.20722606661316]

            rank: Integer representation of the rank of the hashtag relative to others.

                .. code-block:: python

                    tag.rank # >>> 83

            exists: Bool representation of whether or not the passed hashtag exists.

                .. code-block:: python

                    tag.exists # >>> True

            results: List of :py:class:`TagResult` containing information on related hashtags.

                .. code-block:: python

                    tag.results # >>> [Result(tag=..., rank=..., geo=[...,...], media_count=..., relevance=..., abs_relevance=...), ...]
                    result = tag.results[0]

                .. autoclass:: instahashtag.wrapper.tag.TagResult

                    .. autofunction:: instahashtag.wrapper.tag.TagResult.__gt__
        """

        self.hashtag = hashtag
        self.geo = None
        self.rank = None
        self.results = None
        self.exists = None

        if not aio:
            self.data = api.tag(hashtag=self.hashtag)
            self.process()

    async def call(self) -> None:
        """Asynchronously queries the API.

        See note on the top of the :ref:`wrapper` documentation.
        """
        self.data = await api.tag(hashtag=self.hashtag, aio=True)
        self.process()

    def process(self) -> None:
        self.geo = self.data.get("geo", None)
        self.rank = self.data.get("rank", None)
        self.exists = self.data.get("tagExists", None)

        results = self.data.get("results", [])
        if results:
            self.results = [TagResult(**r) for r in results]
        else:
            self.results = results

    def __repr__(self) -> str: # pragma: no cover
        return "Tag(hashtag={}, exists={}, rank={}, results_len={})".format(
            self.hashtag,
            self.exists,
            self.rank,
            len(self.results),
        )

    def __str__(self) -> str: # pragma: no cover
        return self.__repr__()


instahashtag/wrapper/graph.py
from .. import api


class GraphNode:
    def __init__(
        self,
        id: str,
        relevance: float,
        weight: float,
        x: float,
        y: float,
    ) -> None:
        """Object that represents the individual edges inside the ``Graph.nodes`` list.

        Attributes:
            id: Hashtag name.

                .. code-block:: python

                    node.name # >>> liv

            relevance: Relevance of the node.

                .. code-block:: python

                    node.relevance # >>> 0.5417327185029007

            weight: Weight of the node.

                .. code-block:: python

                    node.weight # >>> 0.4523809523809524

            x: Position of the node in relation to the x-axis.

                .. code-block:: python

                    node.x # >>> 0.20431805750484297

            y: Position of the node in relation to the y-axis.

                .. code-block:: python

                    node.y # >>> 0.6194782200730474
        """
        self.id = id
        self.relevance = relevance
        self.weight = weight
        self.x = x
        self.y = y

    def __repr__(self) -> str:  # pragma: no cover
        return "Node(id={}, relevance={}, weight={}, x={}, y={})".format(
            self.id,
            self.relevance,
            self.weight,
            self.x,
            self.y,
        )

    def __str__(self) -> str:  # pragma: no cover
        return self.__repr__()


class GraphEdge:
    def __init__(self, a: str, b: str, id: str, weight: int) -> None:
        """Object that represents the individual edges inside the ``Graph.edges`` list.

        Attributes:
            a: Hashtag name that represents a node in the graph.

                .. code-block:: python

                    edge.a # >>> liv

            b: Hashtag name that represents another node in the graph.

                .. code-block:: python

                    edge.a # >>> thingstodomiami

            id: Order of the connection between nodes.

                .. code-block:: python

                    edge.id # >>> liv#thingstodomiami

            weight: Weight of the edge.

                .. code-block:: python

                    edge.weight # >>> 0.44775669978922017

        Note:
            The edge object of the graph represents `which` node is connected to which other node.
        """
        self.a = a
        self.b = b
        self.id = id
        self.weight = weight

    def __repr__(self) -> str:  # pragma: no cover
        return "Edge(a={}, b={}, id={}, weight={})".format(
            self.a,
            self.b,
            self.id,
            self.weight,
        )

    def __str__(self) -> str:  # pragma: no cover
        return self.__repr__()


class Graph:
    def __init__(self, hashtag: str, aio: bool = False) -> None:
        """Initializes a new Graph object.

        .. code-block:: python

            from instahashtag import Graph

            graph = Graph("miami")

        Attributes:
            hashtag: Hashtag that was used to retrieve information from.

               .. code-block:: python

                    graph.hashtag # >>> miami

            exists: Boolean that dictates whether or not the passed hashtag exists.

               .. code-block:: python

                    graph.exists # >>> True

            root_pos: List of floats indicating where the root position of the graph is.

                .. code-block:: python

                    graph.root_post # >>> [0.4495344797287565, 0.40752168227901403]

            nodes: List of :py:class:`GraphNode` containing information on related hashtags.

                .. code-block:: python

                    graph.nodes # >>> [Node(id=..., relevance=..., weight=..., x=..., y=...), ...]
                    node = graph.nodes[0]

                .. autoclass:: instahashtag.wrapper.graph.GraphNode

            edges: List of :py:class:`GraphEdge` containing information on the connection between hashtags.

                .. code-block:: python

                    graph.edges # >>> [Edge(a=..., b=..., id=...#..., weight=...), ...]
                    edge = graph.edges[0]

                .. autoclass:: instahashtag.wrapper.graph.GraphEdge

        """

        self.hashtag = hashtag
        self.edges = None
        self.exists = None
        self.nodes = None
        self.query = None
        self.root_pos = None

        if not aio:
            self.data = api.graph(hashtag=self.hashtag)
            self.process()

    async def call(self) -> None:
        """Asynchronously queries the API.

        See note on the top of the :ref:`wrapper` documentation.
        """
        self.data = await api.graph(hashtag=self.hashtag, aio=True)
        self.process()

    def process(self) -> None:
        """Processes the incoming data from the API."""

        self.exists = self.data.get("exists", None)
        self.root_pos = self.data.get("root_pos", None)

        edges = self.data.get("edges", [])
        if edges:
            self.edges = [GraphEdge(**e) for e in edges]
        else:
            self.edges = edges

        nodes = self.data.get("nodes", [])
        if nodes:
            self.nodes = [GraphNode(**n) for n in nodes]
        else:
            self.nodes = nodes

    def __repr__(self) -> str:  # pragma: no cover
        return "Graph(hashtag={}, exists={}, root_pos={}, edges_len={}, nodes_len={})".format(
            self.hashtag,
            self.exists,
            self.root_pos,
            len(self.edges),
            len(self.nodes),
        )

    def __str__(self) -> str:  # pragma: no cover
        return self.__repr__()


instahashtag/wrapper/maps.py
from typing import List

from .. import api


class MapsTag:
    """Object that represents the individiual items inside the ``Maps.tags`` list.

    Attributes:
        centroid: List of floats that contains the location of the tag in the map (lon, lat).

            .. code-block::

                tag.centroid # >>> [25.801775593361942, -80.20252247848369]

        tag: Hashtag name.

            .. code-block::

                tag.tag # >>> igersmiami

        weight: Weight of the hashtag on the map.

            .. code-block::

                tag.weight # >>> 49
    """

    def __init__(self, centroid: List[float], tag: str, weight: int) -> None:
        self.centroid = centroid
        self.tag = tag
        self.weight = weight

    def __gt__(self, other: "Result") -> bool:
        """Allows comparisson of :py:class:`MapsTags` objects by their ``weight``.

        Note:
            This function allows the user to utilize the Python built-in functions.

            .. code-block:: python

                from instahashtag import Maps

                # Coordinates designate to Miami, FL.
                maps = Maps(
                    x1=-80.48712034709753,
                    y1=25.750749758162012,
                    x2=-79.82794065959753,
                    y2=25.854604964203453,
                    zoom=12
                )

                min(maps.tags)
                max(maps.tags)
                sort(maps.tags)

        """
        return self.weight > other.weight

    def __repr__(self) -> str:  # pragma: no cover
        return "MapTag(tag={}, centroid={}, weight={})".format(
            self.tag,
            self.centroid,
            self.weight,
        )

    def __str__(self) -> str:  # pragma: no cover
        return self.__repr__()


class Maps:
    def __init__(
        self,
        x1: float,
        y1: float,
        x2: float,
        y2: float,
        zoom: int,
        aio: bool = False,
    ) -> None:
        """Initializes a new map object.

        .. code-block:: python

            from instahashtag import Maps

            # Coordinates designate to Miami, FL.
            maps = Maps(
                x1=-80.48712034709753,
                y1=25.750749758162012,
                x2=-79.82794065959753,
                y2=25.854604964203453,
                zoom=12
            )

        Attributes:
            x1: Top left x-coordinate corner of the map.

                .. code-block:: python

                    maps.x1 # >>> -80.48712034709753

            y1: Top left y-coordinate corner of the map.

                .. code-block:: python

                    maps.y1 # >>> 25.750749758162012

            x2: Bottom right x-coordinate corner of the map.

                .. code-block:: python

                    maps.x2 # >>> -79.82794065959753

            y2: Bottom right y-coordinate corner of the map.

                .. code-block:: python

                    maps.y2 # >>> 25.854604964203453

            zoom: Number from 2 to 16 that designates the zoom factor.

                .. code-block:: python

                    maps.zoom # >>> 12

            count: Number of hashtags in the resulting query.

                .. code-block:: python

                    maps.count # >>> 91

            tags: List of hashtags.

                .. code-block:: python

                    maps.tags # >>> [MapTag(tag=..., centroid=[..., ...], weight=...), ...]
                    tag = maps.tags[0]

                .. autoclass:: instahashtag.wrapper.maps.MapsTag

                    .. autofunction:: instahashtag.wrapper.maps.MapsTag.__gt__

        """

        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2
        self.zoom = zoom

        if not aio:
            self.data = api.maps(
                x1=self.x1,
                y1=self.y1,
                x2=self.x2,
                y2=self.y2,
                zoom=self.zoom,
            )
            self.process()

    async def call(self) -> None:
        """Asynchronously queries the API.

        See note on the top of the :ref:`wrapper` documentation.
        """
        self.data = await api.maps(
            x1=self.x1,
            y1=self.y1,
            x2=self.x2,
            y2=self.y2,
            zoom=self.zoom,
            aio=True,
        )
        self.process()

    def process(self) -> None:
        """Processes the incoming data from the API."""

        self.count = self.data.get("count", None)

        tags = self.data.get("tags", [])
        if tags:
            self.tags = [MapsTag(**t) for t in tags]
        else:
            self.tags = []

    def __repr__(self) -> str:  # pragma: no cover
        return (
            "Graph(x1={}, y1={}, x2={}, y2={}, zoom={}, count={}, tags_len={})".format(
                self.x1,
                self.y1,
                self.x2,
                self.y2,
                self.zoom,
                self.count,
                len(self.tags),
            )
        )

    def __str__(self) -> str:  # pragma: no cover
        return self.__repr__()
